opm/autodiff/BlackoilWellModel.hpp

```diff
@@ -126,44 +115,59 @@ namespace Opm {
 
             // using the solution x to recover the solution xw for wells and applying
             // xw to update Well State
-            void recoverWellSolutionAndUpdateWellState(const BVector& x, WellState& well_state) const;
+            void recoverWellSolutionAndUpdateWellState(const BVector& x);
 
-            int numWells() const;
+            // Check if well equations is converged.
+            bool getWellConvergence(const std::vector<Scalar>& B_avg) const;
 
-            /// return true if wells are available in the reservoir
-            bool wellsActive() const;
+            // return all the wells.
+            const WellCollection& wellCollection() const;
+            WellCollection& wellCollection();
 
-            void setWellsActive(const bool wells_active);
+            // return the internal well state, ignore the passed one.
+            // Used by the legacy code to make it compatible with the legacy well models.
+            const WellState& wellState(const WellState& well_state OPM_UNUSED) const { return wellState(); }
```

Comment:
I'd prefer to see all method implementations (even small ones) in the _impl.hpp file, since that makes it much easier to see where a member is used etc. (need only search one file). Taking that argument further, one can argue in favour of having it all in the hpp file, but for big classes like this it is important to have a readable public interface, so I think we should keep the two files.